components/core/src/clp/StreamingReader.hpp

```diff
@@ -0,0 +1,405 @@
+
+    /**
+     * Frees the current reading buffer.
+     */
+    auto free_reading_buffer() -> void;
+
+    /**
+     * Commits the number of bytes fetched. If the buffer is full, commits it to the fetched buffer
+     * queue.
+     * @param num_bytes_fetched
+     */
+    auto commit_fetching(size_t num_bytes_fetched) -> void;
+
+    /**
+     * Commits the number of bytes read. If the entire buffer is consumed, free the buffer resource.
+     * @param num_bytes_read
+     */
+    auto commit_reading(size_t num_bytes_read) -> void;
+
+    /**
+     * Gets the buffer to fetch (which is the unfilled part of the current fetching buffer). If the
+     * current fetching buffer is full, it will commit the current buffer and get the next buffer
+     * from the buffer pool.
+     * @param fetching_buffer Output the buffer to fetch.
+     * @return true on success, false on failure (transfer aborted).
+     */
+    [[nodiscard]] auto get_buffer_to_fetch(BufferView& fetching_buffer) -> bool;
+
+    /**
+     * Resets the buffer states and transfer manager.
+     * Note:
+     * 1. It is assumed that the thread that calls `reset` would be the reader thread that executed
+     *  `open`.
+     * 2. The transfer should already be terminated (either naturally or aborted) before calling
+     *  this thread.
+     */
+    auto reset() -> void;
+
+    /**
+     * Reads data from the fetched buffers with a given amount of bytes.
+     * @param num_bytes_to_read
+     * @param num_bytes_read Returns the number of bytes read.
+     * @param dst The pointer to the destination buffer. If the buffer is not nullptr, copy the data
+     * to the destination.
+     * @return ErrorCode_EndOfFile if the buffer doesn't contain any more data.
+     * @return ErrorCode_Success on success.
+     */
+    [[nodiscard]] auto
+    read_from_fetched_buffers(size_t num_bytes_to_read, size_t& num_bytes_read, char* dst)
+            -> ErrorCode;
+
+    std::string m_src_url;
+    size_t m_file_pos{0};
+
+    size_t m_buffer_pool_size;
+    size_t m_buffer_size;
+    size_t m_num_fetched_buffer{0};
+    size_t m_curr_fetching_buffer_idx{0};
+    size_t m_fetching_buffer_pos{0};
+
+    uint32_t m_connection_timeout{cDefaultConnectionTimeout};
+    uint32_t m_operation_timeout{cDefaultOperationTimeout};
+
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
+    std::vector<std::unique_ptr<char[]>> m_buffer_pool;
+    std::queue<BufferView> m_fetched_buffer_queue;
+    std::optional<BufferView> m_fetching_buffer{std::nullopt};
+    std::optional<BufferView> m_reading_buffer{std::nullopt};
+
+    std::mutex m_buffer_resource_mutex;
+    std::condition_variable m_cv_fetcher;
+    std::condition_variable m_cv_reader;
+
+    std::unique_ptr<std::thread> m_transfer_thread{nullptr};
+    std::atomic<bool> m_transfer_aborted{false};
+    std::atomic<bool> m_transfer_terminated{false};
+    StatusCode m_status_code{StatusCode::NotInit};
```

Comment:
From the current implementation, it is fine to not use `std::atomic` as long as `m_transfer_terminated` is atomic. But I agree with the comments since it would reduce the chance of introducing new synchronization bugs in the future devs.